From 9ea7557a2fd1cfcd7239b792841329856e0d74a2 Mon Sep 17 00:00:00 2001
From: Mikolaj Klikowicz <mklikowicz@antmicro.com>
Date: Fri, 13 Feb 2026 16:47:24 +0100
Subject: [PATCH 30/37] instrumentation: Add recursive disable of functions

Added support for recursively disabling certain functions from emitting
instrumentation events.

Signed-off-by: Mikolaj Klikowicz <mklikowicz@antmicro.com>
---
 .../zephyr/instrumentation/instrumentation.h  |  6 ++
 subsys/instrumentation/CMakeLists.txt         | 11 +++
 subsys/instrumentation/Kconfig                | 11 +++
 subsys/instrumentation/common/instr_common.c  | 79 +++++++++++++++++++
 .../handlers/instr_handlers_gcc.c             | 14 +++-
 subsys/instrumentation/instr_exclude_gen.py   | 23 ++++++
 6 files changed, 142 insertions(+), 2 deletions(-)
 create mode 100755 subsys/instrumentation/instr_exclude_gen.py

diff --git a/include/zephyr/instrumentation/instrumentation.h b/include/zephyr/instrumentation/instrumentation.h
index c989d8c4f83..888c5508915 100644
--- a/include/zephyr/instrumentation/instrumentation.h
+++ b/include/zephyr/instrumentation/instrumentation.h
@@ -232,6 +232,12 @@ void *instr_get_trigger_func(void);
  */
 void *instr_get_stop_func(void);
 
+void instr_recursive_exclude_try_enable(void *callee);
+
+void instr_recursive_exclude_try_disable(void *callee);
+
+bool instr_recursive_exclude_enabled(void);
+
 #ifdef __cplusplus
 }
 #endif
diff --git a/subsys/instrumentation/CMakeLists.txt b/subsys/instrumentation/CMakeLists.txt
index a711c3d5ab7..9a79d245e3d 100644
--- a/subsys/instrumentation/CMakeLists.txt
+++ b/subsys/instrumentation/CMakeLists.txt
@@ -7,6 +7,17 @@ if(CONFIG_INSTRUMENTATION)
 
   zephyr_library()
 
+  set(instr_recursive_exclude ${CMAKE_BINARY_DIR}/instr_recursive_exclude.c)
+
+  add_custom_command(
+    OUTPUT ${instr_recursive_exclude}
+    COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/instr_exclude_gen.py ${CONFIG_INSTRUMENTATION_RECURSIVE_EXCLUDE_LIST} ${instr_recursive_exclude}
+    DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/instr_exclude_gen.py
+    VERBATIM
+  )
+
+  zephyr_library_sources_ifdef(CONFIG_INSTRUMENTATION_RECURSIVE_EXCLUDE ${instr_recursive_exclude})
+
   zephyr_library_sources(common/instr_common.c)
   zephyr_library_sources(timestamp/timestamp.c)
 
diff --git a/subsys/instrumentation/Kconfig b/subsys/instrumentation/Kconfig
index 9545834f933..0309c7b93bf 100644
--- a/subsys/instrumentation/Kconfig
+++ b/subsys/instrumentation/Kconfig
@@ -24,6 +24,16 @@ endchoice
 rsource "transport/Kconfig.uart"
 rsource "transport/Kconfig.tracing"
 
+config INSTRUMENTATION_RECURSIVE_EXCLUDE
+	bool "Recursively exclude children of specified functions from instrumentation"
+	default y
+
+if INSTRUMENTATION_RECURSIVE_EXCLUDE
+config INSTRUMENTATION_RECURSIVE_EXCLUDE_LIST
+	string "Comma separated list of functions"
+	default "sys_clock_isr,z_impl_k_sem_give,z_impl_k_sem_take,z_impl_k_mutex_lock,z_impl_k_mutex_unlock,z_impl_k_sleep,printk,instr_timestamp_ns"
+endif
+
 config INSTRUMENTATION_MODE_CALLGRAPH
 	bool "Call graph mode (Tracing)"
 	default y
@@ -104,6 +114,7 @@ config INSTRUMENTATION_EXCLUDE_FUNCTION_LIST
 
 config INSTRUMENTATION_EXCLUDE_FILE_LIST
 	string "Exclude file list"
+	default "include/zephyr/spinlock.h,include/zephyr/timing/timing.h,zephyr/subsys/tracing/"
 	depends on INSTRUMENTATION_MODE_CALLGRAPH || INSTRUMENTATION_MODE_STATISTICAL
 	help
 	  Set the list of files that are excluded from instrumentation. The
diff --git a/subsys/instrumentation/common/instr_common.c b/subsys/instrumentation/common/instr_common.c
index 8bedced932e..b039695cfe5 100644
--- a/subsys/instrumentation/common/instr_common.c
+++ b/subsys/instrumentation/common/instr_common.c
@@ -26,6 +26,85 @@
 extern void z_thread_mark_switched_in(void);
 extern void z_thread_mark_switched_out(void);
 
+#if defined(CONFIG_INSTRUMENTATION_RECURSIVE_EXCLUDE)
+/* Generated by instr_exclude_gen.py */
+extern unsigned int instr_exclude_list_len;
+extern void (*instr_exclude_list[])(void);
+
+__no_instrumentation__ static int instr_exclude_list_cmp(const void *a_ptr, const void *b_ptr)
+{
+	uintptr_t a = *(uintptr_t *)a_ptr;
+	uintptr_t b = *(uintptr_t *)b_ptr;
+
+	if (a == b) {
+		return 0;
+	} else if (a > b) {
+		return -1;
+	} else {
+		return 1;
+	}
+}
+
+__no_instrumentation__ static int instr_recursive_exclude_init(void)
+{
+	qsort(instr_exclude_list, instr_exclude_list_len, sizeof(instr_exclude_list[0]),
+	      instr_exclude_list_cmp);
+	return 0;
+}
+
+__no_instrumentation__ static bool instr_recursive_exclude_find(void *p)
+{
+	return bsearch(&p, instr_exclude_list, instr_exclude_list_len,
+		       sizeof(instr_exclude_list[0]), instr_exclude_list_cmp) != NULL;
+}
+
+SYS_INIT(instr_recursive_exclude_init, APPLICATION, 0);
+
+static Z_THREAD_LOCAL bool _instr_recursive_exclude_enabled;
+static Z_THREAD_LOCAL void *_instr_recursive_exclude_func;
+
+__no_instrumentation__ void instr_recursive_exclude_try_enable(void *callee)
+{
+	if (instr_recursive_exclude_enabled()) {
+		return;
+	}
+
+	if (instr_recursive_exclude_find(callee)) {
+		_instr_recursive_exclude_enabled = true;
+		_instr_recursive_exclude_func = callee;
+	}
+}
+
+__no_instrumentation__ void instr_recursive_exclude_try_disable(void *callee)
+{
+	if (_instr_recursive_exclude_func == callee) {
+		_instr_recursive_exclude_enabled = false;
+		_instr_recursive_exclude_func = NULL;
+	}
+}
+
+__no_instrumentation__ bool instr_recursive_exclude_enabled(void)
+{
+	return _instr_recursive_exclude_enabled;
+}
+
+#else
+
+__no_instrumentation__ void instr_recursive_exclude_try_enable(void *callee)
+{
+}
+
+__no_instrumentation__ void instr_recursive_exclude_try_disable(void *callee)
+{
+}
+
+__no_instrumentation__ bool instr_recursive_exclude_enabled(void)
+{
+	return false;
+}
+
+#endif
+
 /*
  * Memory buffer to store instrumentation event records has the following modes:
  *
diff --git a/subsys/instrumentation/handlers/instr_handlers_gcc.c b/subsys/instrumentation/handlers/instr_handlers_gcc.c
index 8251a334ed6..f57801f3137 100644
--- a/subsys/instrumentation/handlers/instr_handlers_gcc.c
+++ b/subsys/instrumentation/handlers/instr_handlers_gcc.c
@@ -1,5 +1,7 @@
 /*
  * Copyright 2023 Linaro
+ * Copyright (c) 2026 Antmicro
+ * Copyright (c) 2026 Analog Devices
  *
  * SPDX-License-Identifier: Apache-2.0
  */
@@ -33,7 +35,11 @@ void __cyg_profile_func_enter(void *callee, void *caller)
 		return;
 	}
 
-	instr_event_handler(INSTR_EVENT_ENTRY, callee, caller);
+	if (!instr_recursive_exclude_enabled()) {
+		instr_event_handler(INSTR_EVENT_ENTRY, callee, caller);
+	}
+
+	instr_recursive_exclude_try_enable(callee);
 }
 
 __no_instrumentation__
@@ -47,7 +53,11 @@ void __cyg_profile_func_exit(void *callee, void *caller)
 		return;
 	}
 
-	instr_event_handler(INSTR_EVENT_EXIT, callee, caller);
+	instr_recursive_exclude_try_disable(callee);
+
+	if (!instr_recursive_exclude_enabled()) {
+		instr_event_handler(INSTR_EVENT_EXIT, callee, caller);
+	}
 
 	/* Turn off instrumentation if stopper returns */
 	if (callee == instr_get_stop_func() && instr_turned_on()) {
diff --git a/subsys/instrumentation/instr_exclude_gen.py b/subsys/instrumentation/instr_exclude_gen.py
new file mode 100755
index 00000000000..019a7cd7431
--- /dev/null
+++ b/subsys/instrumentation/instr_exclude_gen.py
@@ -0,0 +1,23 @@
+# Copyright (c) 2026 Antmicro
+# Copyright (c) 2026 Analog Devices
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
+
+f_str = sys.argv[1] if len(sys.argv) >= 2 else ""
+
+fs = f_str.split(",") if f_str else []
+
+out_path = sys.argv[2]
+
+with open(out_path, "w") as out_f:
+    for f in fs:
+        print(f"extern void {f}(void);", file=out_f)
+
+    print(f"unsigned instr_exclude_list_len = {len(fs)};", file=out_f)
+    print("void (*instr_exclude_list[])(void) __attribute__((retain)) = {", file=out_f)
+
+    for f in fs:
+        print(f"    {f},", file=out_f)
+
+    print("};", file=out_f)
-- 
2.47.3

